Rubberduck.Parsing/VBA/RubberduckParserReimpl.cs

```diff
@@ -0,0 +1,356 @@
+        private readonly Dictionary<VBComponent, ITokenStream> _tokenStreams = new Dictionary<VBComponent, ITokenStream>();
+        private readonly Dictionary<VBComponent, IList<CommentNode>> _comments = new Dictionary<VBComponent, IList<CommentNode>>();
+        private readonly IDictionary<VBComponent, IDictionary<Tuple<string, DeclarationType>, Attributes>> _componentAttributes
+            = new Dictionary<VBComponent, IDictionary<Tuple<string, DeclarationType>, Attributes>>();
+
+
+        private readonly ReferencedDeclarationsCollector _comReflector;
+
+        private readonly VBE _vbe;
+        private readonly RubberduckParserState _state;
+        private readonly IAttributeParser _attributeParser;
+
+        public RubberduckParserReimpl(VBE vbe, RubberduckParserState state, IAttributeParser attributeParser)
+        {
+            _vbe = vbe;
+            _state = state;
+            _attributeParser = attributeParser;
+
+            _comReflector = new ReferencedDeclarationsCollector();
+
+            state.ParseRequest += ReparseRequested;
+            state.StateChanged += StateOnStateChanged;
+        }
+
+        private void StateOnStateChanged(object sender, EventArgs e)
+        {
+            var args = e as ParserStateEventArgs;
+            if (args.State == ParserState.Parsed)
+            {
+                Task.Run(() => Resolve(_central.Token));
+                // Resolving should be triggered.. not our job?
+            }
+        }
+
+        private void ReparseRequested(object sender, EventArgs e)
+        {
+            var args = e as ParseRequestEventArgs;
+            if (args.IsFullReparseRequest)
+            {
+                Cancel();
+                ParseAll();
+            }
+            else
+            {
+                Cancel(args.Component);
+                ParseAsync(args.Component, CancellationToken.None, _state.GetRewriter(args.Component));
+            }
+        }
+
+        /// <summary>
+        /// Starts parsing all components of all unprotected VBProjects associated with the VBE-Instance passed to the constructor of this parser instance.
+        /// </summary>
+        private void ParseAll()
+        {
+            var projects = _vbe.VBProjects
+                .Cast<VBProject>()
+                .Where(project => project.Protection == vbext_ProjectProtection.vbext_pp_none);
+
+            var components = projects.SelectMany(p => p.VBComponents.Cast<VBComponent>());
+            // invalidation cleanup should go into ParseAsync?
+            foreach (var invalidated in _componentAttributes.Keys.Except(components))
+            {
+                _componentAttributes.Remove(invalidated);
+            }
+
+            foreach (var vbComponent in components)
+            {
+                while (!_state.ClearDeclarations(vbComponent)) { }
```

Comment:
This is questionable. Can this be resolved in an asynchronous manner, perhaps using events? Otherwise a rewrite of what `_state.ClearDeclarations()` really does? Right now it's just eating CPU time -- assuming it's not optimized away.